client/session.go

```diff
@@ -563,3 +595,943 @@ func (s *session) Close() error {
+					continue
+				}
+
+				blocks[i].size = *b.Size
+			}
+			ch <- blocksMetadata{
+				peer:   peer,
+				id:     elem.ID,
+				blocks: blocks,
+			}
+		}
+
+		return nil
+	}
+	for moreResults {
+		if err := retrier.Attempt(attemptFn); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *session) streamBlocksFromPeers(
+	shard uint32,
+	peers []hostQueue,
+	ch <-chan blocksMetadata,
+	result *blocksResult,
+) error {
+	var (
+		retrier = xretry.NewRetrier(xretry.NewOptions().
+			BackoffFactor(2).
+			Max(3).
+			InitialBackoff(time.Second).
+			Jitter(true))
+		enqueueCh           = newEnqueueChannel()
+		peerBlocksBatchSize = s.streamBlocksBatchSize
+	)
+
+	// Consume the incoming metadata and enqueue to the ready channel
+	go func() {
+		s.streamCollectedBlocksMetadata(len(peers), ch, enqueueCh)
+		// Begin assessing the queue and how much is processed, once queue
+		// is entirely processed then we can close the enqueue channel
+		enqueueCh.closeOnAllProcessed()
+	}()
+
+	// Fetch blocks from peers as results become ready
+	peerQueues := make(peerBlocksQueues, 0, len(peers))
+	for _, peer := range peers {
+		peer := peer
+		size := peerBlocksBatchSize
+		workers := s.streamBlocksWorkers
+		drainEvery := 100 * time.Millisecond
+		queue := newPeerBlocksQueue(peer, size, drainEvery, workers, func(batch []*blocksMetadata) {
+			s.streamBlocksBatchFromPeer(retrier, peer, batch, result, enqueueCh)
+		})
+		peerQueues = append(peerQueues, queue)
+	}
+
+	var (
+		currStart, currEligible []*blocksMetadata
+		blocksMetadataQueues    []blocksMetadataQueue
+	)
+	for perPeerBlocksMetadata := range enqueueCh.get() {
+		// Filter and select which blocks to retrieve from which peers
+		if err := s.selectBlocksForSeriesFromPeerBlocksMetadata(
+			perPeerBlocksMetadata, peerQueues,
+			currStart, currEligible, blocksMetadataQueues,
+		); err != nil {
+			enqueueCh.trackProcessed(1)
+			continue
+		}
+
+		// Insert work into peer queues
+		var wg sync.WaitGroup
+		for _, peerBlocksMetadata := range perPeerBlocksMetadata {
+			if len(peerBlocksMetadata.blocks) == 0 {
+				// No blocks assigned
+				continue
+			}
+			queue := peerQueues.findQueue(peerBlocksMetadata.peer)
+			wg.Add(1)
+			queue.enqueue(peerBlocksMetadata, wg.Done)
+		}
+		// Launch asynchronous waiter to mark completion
+		go func() {
+			wg.Wait()
+			enqueueCh.trackProcessed(1)
+		}()
+	}
+
+	// Close all queues
+	peerQueues.closeAll()
+
+	return nil
+}
+
+func (s *session) streamCollectedBlocksMetadata(
+	peersLen int,
+	ch <-chan blocksMetadata,
+	enqueueCh *enqueueChannel,
+) {
+	metadata := make(map[string]*receivedBlocks)
+
+	// Receive off of metadata channel
+	for {
+		m, ok := <-ch
+		if !ok {
+			break
+		}
+
+		received, ok := metadata[m.id]
+		if !ok {
+			received = &receivedBlocks{
+				results: make([]*blocksMetadata, 0, peersLen),
+			}
+			metadata[m.id] = received
+		}
+		if received.submitted {
+			// Already submitted to enqueue channel
+			continue
```

Comment:
This will happen say if the pagination has a bug in it and your read the same ID twice for some reason or potentially if you start reading metadata off a new host during the streaming process (currently not supported or possible).